[gavingoh99] iP

gavingoh99 opened the pull request:
# Mona 👍 
> “Everyone needs a good friend.” – Gavin Goh 

And Mona can be just the friend you need! She's

- friendly
- text-based
- a ~~procrastination~~ *PRODUCTIVITY* enabler


## How do I get started?
All you need to do is, 
1. download it from [here](https://github.com/gavingoh99/ip/releases/tag/A-Jar).
1. double-click it.
1. add your tasks.
1. let Mona manage your tasks for you 😉
```
java --jar mona.jar
```
That's it! It's that **simple**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

gavingoh99 committed:
modified gitignore to disregard class files

gavingoh99 committed:
added features for level 1

gavingoh99 committed:
added features for level 1 corrected

gavingoh99 committed:
removed Duke template

gavingoh99 committed:
added features for level 2

gavingoh99 committed:
added features for level 3

gavingoh99 committed:
added features for level 4

gavingoh99 committed:
added textuitesting

gavingoh99 committed:
added exception and handling of errors for level 5

gavingoh99 committed:
added delete feature

gavingoh99 committed:
feat: added logging functionality

gavingoh99 committed:
Merge branch 'branch-Level-7'

gavingoh99 committed:
misc: git ignore chatbot log

gavingoh99 committed:
feat: added datetime functionality

gavingoh99 committed:
Merge branch 'branch-Level-8'

gavingoh99 committed:
refactor: added new classes to abstract functionality

gavingoh99 committed:
refactor: abstracted task list and storage functions

gavingoh99 committed:
refactor: abstracted parser and ui functionality

gavingoh99 committed:
Merge branch 'A-MoreOOP'

gavingoh99 committed:
refactor: transferred files into Mona package

gavingoh99 committed:
Merge branch 'A-Packages'

gavingoh99 committed:
Merge branch 'add-gradle-support'

gavingoh99 committed:
feat: integrated gradle functionality

gavingoh99 committed:
feat: added unit testing for parser and tasklist

gavingoh99 committed:
Merge branch 'A-JUnit'

gavingoh99 committed:
feat: added gradle build function

gavingoh99 committed:
Merge branch 'A-Jar'

gavingoh99 committed:
docs: added javadocs

gavingoh99 committed:
misc: modified code to adhere to styling conventions

gavingoh99 committed:
feat: added find feature and fixed bug regarding IOException being thrown

gavingoh99 committed:
Merge branch 'A-JavaDoc'

gavingoh99 committed:
Merge branch 'master' into A-CodingStandard

gavingoh99 committed:
Merge branch 'A-CodingStandard'

gavingoh99 committed:
misc: resolve merge conflict

gavingoh99 committed:
Merge branch 'Level-9'

gavingoh99 committed:
feat: added checkstyle capabilities and integrated functionality with gradle

gavingoh99 committed:
Merge branch 'A-CheckStyle'

gavingoh99 committed:
feat: ported cli to gui and updated ui to fit dimensions

gavingoh99 committed:
Merge branch 'Level-10'

yashma-sonara commented on src/main/java/Mona/Task.java:
```diff
@@ -0,0 +1,36 @@
+package mona;
+public class Task {
```
Since your task types inherit from Task class, perhaps you could consider making Task into its own package?

yashma-sonara commented on src/main/java/Mona/Storage.java:
```diff
@@ -0,0 +1,96 @@
+     * Method for updating the logs when the task list is modified while Mona is running
+     * @param currentTasks Mona's task list which the user can directly interact with
+     */
+    public void writeToFile(List<Task> currentTasks) {
+        File log = new File(this.filePath);
+        StringBuilder sb = new StringBuilder();
+        for (Task task : currentTasks) {
+            sb.append(task.parseToLogRepresentation() + "\n");
+        }
+        try {
+            FileWriter fw = new FileWriter(log.getAbsoluteFile());
+            BufferedWriter bw = new BufferedWriter(fw);
+            bw.write(sb.toString());
+            bw.close();
+        } catch (IOException e) {
+            System.out.println("Problem writing to log: " + e.getMessage());
+        }
+    }
```
I like how you are using StringBuilder and BufferedWriter to improve performance and efficiency.

yashma-sonara commented on src/main/java/Mona/Parser.java:
```diff
@@ -0,0 +1,68 @@
+        case TODO:
+            details[0] = input.substring(5);
+            break;
+        case DEADLINE:
+            String rest = input.substring(9);
+            String[] parts = rest.split(" /by ");
+            details[0] = parts[0];
+            details[1] = parts[1];
```
I was wondering is the case where a deadline command is entered (ex. deadline read 2024-03-04) Where the "/by" is missing is being handled? Same for event times?

yashma-sonara commented on src/main/java/Mona/Todo.java:
```diff
@@ -0,0 +1,16 @@
+package mona;
```
Maybe you could add a class header for javadocs?

yashma-sonara reviewed (commented):
The code LGTM. Just a few nits to look at. Good work 👍

yeoshuheng commented on src/main/java/Mona/Deadline.java:
```diff
@@ -0,0 +1,25 @@
+public class Deadline extends Task {
+
+    protected LocalDateTime by;
+    protected DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("d 'of' MMMM yyyy, ha");
```
Great use of DateTimeFormatter here! Makes your code handle the dates a lot cleaner :)

yeoshuheng commented on src/main/java/Mona/Ui.java:
```diff
@@ -0,0 +1,75 @@
+/**
+ * This class contains all logic regarding Mona's response
+ */
+public class Ui {
```
Great representation / abstraction of the Ui!

Nicely done.

yeoshuheng commented on src/main/resources/images/Mona.jpeg:
cute :)

yeoshuheng commented on src/test/java/Mona/ParserTest.java:
```diff
@@ -0,0 +1,19 @@
+class ParserTest {
+
+    @Test
+    public void testGetCommand() throws MonaException {
```
Maybe it could be better to extend this test to various forms of invalid inputs too, such as event commands without a "/from" and "/to", this could possibly lead to a greater coverage of the Parser class.

yeoshuheng commented on src/main/java/Mona/Storage.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /**
+     * Method for translating the tasks stored in the logs into their corresponding Tasks
```
Rather than returning null, would it be better to throw some form of ParserException here?

yeoshuheng reviewed (commented):
Great work.
Code is clean, easy to read and understand and well structured.

LGTM.

gavingoh99 committed:
Mona: add assertions to methods

A switch statement in Mona currently does not use assertions.

Let's add assertions to the switch statement and the block after
to check against our assumption that currCommand can only take
the finite values that have been specified in the enumeration as well
as to ensure that execution in the getResponse() method does not
reach past the switch statement.

Assertions are statements to test assumptions in the program and
are used here in accordance to Java's specifications on programming
with Assertions to test for internal invariants.

gavingoh99 committed:
style: fix for checkstyle

gavingoh99 committed:
Storage readLog(): Replace iterative implementation with Streams

readLog() utilises iteration to check through each line of the logs
and parses it into a relevant task before adding it to the task list.

Let's replace the iteration with the Streams API to pull the lines out
from the logs and map each line using relevant helper methods.

Java's Stream API represent a more functional programming way of
handling the same task which tends to be more descriptive for
readability and also encourages less mutability.

gavingoh99 committed:
feat: added new update command

gavingoh99 committed:
Merge pull request #1 from gavingoh99/A-Assertions

Mona: add assertions to methods

gavingoh99 committed:
Merge branch 'master' into A-CodeQuality

gavingoh99 committed:
Merge pull request #2 from gavingoh99/A-CodeQuality

Improvements to Code Quality based off A-CodeQuality

gavingoh99 committed:
misc: resolved merge conflict

gavingoh99 committed:
Merge pull request #3 from gavingoh99/A-Streams

Integration of Streams based on A-Streams branch

gavingoh99 committed:
misc: resolved merge conflict

gavingoh99 committed:
Merge pull request #4 from gavingoh99/BCD-Extension

Implementation of C-Update Feature based off of BCD-Extension task

gavingoh99 committed:
style: fix for checktstyle

gavingoh99 committed:
perf: implement CI

gavingoh99 committed:
fix: unit testing logic

gavingoh99 committed:
Merge pull request #5 from gavingoh99/A-CI

Introduce CI/CD using Github Actions according to A-CI task

shavonneg commented on src/main/java/Mona/MainWindow.java:
```diff
@@ -0,0 +1,51 @@
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+/**
+ * Controller for MainWindow. Provides the layout for the other controls.
+ */
```
Nice that you added javadocs for your classes! Makes things really comprehensive for people to understand how your code interacts with one another.

shavonneg commented on src/main/java/Mona/MainWindow.java:
```diff
@@ -0,0 +1,51 @@
+     * the dialog container. Clears the user input after processing.
+     */
+    @FXML
+    private void handleUserInput() {
```
Good job with the naming of your methods, clear use of verbs and camelCase to adhere to the coding standards.

shavonneg commented on src/main/java/Mona/Mona.java:
```diff
@@ -0,0 +1,88 @@
+package mona;
```
Perhaps you can divide your classes into multiple packages to accommodate the larger number of classes?

shavonneg commented on src/main/java/Mona/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+     */
+    public String displayList() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("  ______________________________________\n");
```
Perhaps you can let a method handle the printing of lines for Ui purposes?

shavonneg reviewed (commented):
Overall LGTM! Your code is easy to read and no glaring coding standard violations were spotted. Since there are quite a few classes you can consider splitting it into more packages.

gavingoh99 committed:
feat: added greeting to GUI and modified messages to be more unique

gavingoh99 committed:
refactor: modified logic in monda to better adhere to SLAP

gavingoh99 committed:
refactor: modified logic in parser for exception handling and adherence to SLAP

gavingoh99 committed:
style: added javadocs for new methods and modificiations to adhere to checkstyle

gavingoh99 committed:
feat: included assertion for update enum for getCommand unit test

gavingoh99 committed:
Merge pull request #7 from gavingoh99/branch-FinalTweaks

Branch final tweaks

gavingoh99 committed:
docs: add ui.png and user guide for mona

gavingoh99 committed:
Merge pull request #8 from gavingoh99/A-UserGuide

docs: add ui.png and user guide for mona

gavingoh99 committed:
docs: update repo readme and config.yml

gavingoh99 committed:
docs: modify readme to be more interpretable on github pages
